쥬스 메이커 [STEP 3] Judy, Finnn

finnn1 opened the pull request:
담당 리뷰어 : @corykim0829
안녕하세요 Cory! 쥬스 메이커 STEP3 PR 보내드립니다 ☺️
이번에도 잘 부탁드립니다~! 🙏

---

# [STEP 3] 기능구현 및 코드설명 🧑‍🏫
- `FruitStockViewController` : 재고 추가 화면의 컨트롤러
-> `fruitStore` : 뷰 컨트롤러에서 넘겨받은 `FruitStore` 인스턴스
-> `setStepper` : stepper의 기본 value를 설정하는 함수
-> `setStockLabel` : 재고 lable에 텍스트를 설정하는 함수
-> `closeButtonTapped` : 버튼을 탭할 경우 현재 화면을 닫는 함수
-> `stepperTapped` : 스태퍼를 탭할 경우 발생하는 액션 함수
- `FruitStore`
-> `modifyStock` : 남은 재고를 원하는 숫자로 수정하는 함수
-> `stockAsDouble` : 남은 재고를 Double 타입으로 반환하는 함수
- `ViewControlelr`
-> `prepare` : 세그웨이가 작동할 경우 동작하는 함수.
- `DoubleExtension` : Double 타입을 String으로 반환해주는 함수


<br/>

# [STEP 3] 고민한점 및 해결방안 🤔
## 고민한점

### 1. 다른 ViewController로 데이터를 전달하는 방법
데이터 전달에 관해 고민하는 과정에서 화면 이동방법에 따라 데이터를 전달할 수 있는 방법이 달라진다는 것을 배우게 되었습니다. '재고 추가' 버튼을 누를 경우에는 직접세그웨이 방식을 사용했기 때문에 `prepare` 메서드를 오버라이드 하여 데이터를 전달했습니다. 재고 품절시에 버튼을 누르면 이동하는 경우는 `Storyboard`의 `identifier`를 활용해 `instantiateViewController` 메서드로 인스턴스를 받아서 화면이동을 했기 때문에, 해당 인스턴스에 직접 접근해서 데이터를 전달하는 방식을 사용해 해결할 수 있었습니다.

<br/>

### 2. Extenstion 활용
여러개의 레이블에 재고를 업데이트해주는 과정에서 `Double` 형을 `String`으로 변경해주어야 하는 작업을 반복적으로 하게 되었습니다. 이 부분을 `Double`에 `extension`을 추가해 `toString` 함수를 구현함으로써 가독성과 코드의 유연성을 향상시킬 수 있었습니다.

<br/>

### 3. 화면이동 방식 - 네비게이션
처음에는 `alert`와 같이 `present` 메서드를 활용해 화면을 이동하려 했으나, 이렇게 할 경우 네이게이션바를 활용할 수 없는 상황이 발생해 네비게이션 방식으로 화면을 이동하는 함수인 `pushViewController` 및 `popViewController` 메서드를 사용해 해결했습니다.

<br/>

### 4. Stepper를 이용한 재고 수정
#### 1) tag를 이용한 Stepper 분류
STEP 2에서 쥬스 주문 버튼들을 `tag`로 분류했던 방식과 동일하게 `Fruit` 열거형에 `Int`로 1부터 `rawValue`를 지정하여 Stepper의 `tag`로 분류했습니다. 덕분에 하나의 액션함수로 모든 과일의 재고를 수정할 수 있었습니다.

#### 2) 재고 변경 반영하기
수정한 재고를 변경하는 방식으로 두 가지를 고민 했습니다.
> 1. Stepper를 누를 때마다 재고를 변경하기
> 2. 모든 변경이 끝나고 이전 화면으로 되돌아갈 때 한 번에 변경하기

<br/>

Stepper를 누를 때마다 변경이 발생하면 비효율적일까 싶어 2번 방식으로 도전해봤습니다. 하지만 재고가 변경되지 않은 과일도 일괄적으로 수정되어야 했고, 과일별로 Stepper의 `value`를 불러오는게 까다로워 1번 방식을 선택하게 되었습니다.

<br/>

# [STEP 3] 배운개념 💡
```swift
navigationItem.hidesBackButton = true
```
- 네비게이션으로 화면 전환 시 자동으로 생성되는 뒤로가기 버튼을 숨길 수 있다.
```swift
String(format: "%.0f", value)
```
- String으로 전환 시 지정한 형식에 맞게 변경된다.

```swift
navigationController?.popViewController(animated: true)
```
- 이전 네비게이션 화면으로 되돌아간다. 
---

## 궁금한 점 🥲
### 1.  레이블과 스태퍼 설정 간소화

어떠한 `Button`을 눌러 이벤트가 발생한 경우에는, `@IBAction` 함수의 파라미터로 들어오는 `sender`의 `tag`를 활용해서 구분이 가능하기 때문에 버튼 각각의 프로퍼티를 만들어주지 않아도 어떤 버튼이 눌렸는지 구분이 가능해, 코드를 깔끔하게 작성하는 것이 가능했습니다. ⬇️
```swift
@IBAction func orderJuiceButtonTapped(sender: UIButton) {
    // sender.tag 를 사용해 어떤 버튼이 눌렸는지 확인 가능 ⬇️
    guard let juiceMenu = Menu(rawValue: sender.tag) else { return }
    
    do{
        try juiceMaker.make(juiceMenu: juiceMenu)
    } catch JuiceMakerError.outOfStock {
        showOutOfStockAlert()
    } catch {
        print("알 수 없는 에러입니다.")
    }
    
    showCompleteAlert(juice: juiceMenu)
    updateStock()
}
```

<br/>

하지만 아래의 코드에서 `Stepper`나 `Label`과 같은 `@IBOutlet` 객체는 일일이 선언해주지 않으면 해당 값을 수정하는 것이 불가능해서, `stockAsDouble(fruit:)`나 `stockAsString(fruit:)` 처럼 동일한 작업을 하는 메서드임에도 각각의 `Stepper` 및 `Label`에 반복적으로 사용할 수 밖에 없었습니다. ⬇️
```swift
private func setStepper() {
    // stockAsDouble(fruit:) 메서드가 중복사용되고 있음 ⬇️
    strawberryStepper.value = fruitStore.stockAsDouble(fruit: .strawberry)
    bananaStepper.value = fruitStore.stockAsDouble(fruit: .banana)
    pineappleStepper.value = fruitStore.stockAsDouble(fruit: .pineapple)
    kiwiStepper.value = fruitStore.stockAsDouble(fruit: .kiwi)
    mangoStepper.value = fruitStore.stockAsDouble(fruit: .mango)
}

private func setStockLabel() {
    // stockAsString(fruit:) 메서드가 중복사용되고 있음 ⬇️
    strawberryStockLabel.text = fruitStore.stockAsString(fruit: .strawberry)
    bananaStockLabel.text = fruitStore.stockAsString(fruit: .banana)
    pineappleStockLabel.text = fruitStore.stockAsString(fruit: .pineapple)
    kiwiStockLabel.text = fruitStore.stockAsString(fruit: .kiwi)
    mangoStockLabel.text = fruitStore.stockAsString(fruit: .mango)
}
```

중복되는 코드를 줄이고자 노력해보았는데 결국 답을 찾지 못해서, 혹시 위와 같은 경우에 저희가 미처 생각치 못한 더 나은 방법이 있지 않을지 궁금해서 질문드립니다 🙏🙏
 
<br/>

### 2. FruitStore를 사용한 재고 변경
재고 추가 화면인 `FruitStockViewController`는 메인화면인 `ViewController`에서 `JuiceMaker`의 프로퍼티인 `FruitStore`의 인스턴스를 전달받아 재고를 변경하게 했습니다. 

```swift
struct JuiceMaker {
    let fruitStore = FruitStore(quantity: 10)
    ...
}
```

`FrultStore` 클래스에 재고를 변경하는 `modifyStock` 메서드를 추가해서 직접 변경할 수 있었습니다. 
```swift
class FruitStore {
    func modifyStock(fruit: Fruit, to number: Int) {
        fruitStock[fruit] = number
    }
    ...
}
```

처음에는 `ViewController`에서 `JuiceMaker`를 이용해서 쥬스를 만들고, 재고 수량을 가져온 것 처럼 `FruitStockViewController`에서도 `JuiceMaker`를 넘겨받아 사용하려 했습니다. 하지만 결국 `JuiceMaker`에 `FruitStore`의 메서드를 사용하는 메서드를 추가해야 했고 비효율적이라 여겨 직접 `FruitStore`를 이용해서 재고 변경을 하게 됐습니다.

코드는 잘 작동하지만 의미상 `JuiceMaker`를 거치지 않고 직접적으로 재고 변경을 해도 괜찮은 것인지, 만약 그렇다고 하면 `JuiceMaker`는 쥬스를 만들기만 하는 용도이고 쥬스의 재고 관리는 `FruitStore`가 하게 되는건지 궁금합니다...🥲

finnn1 committed:
feat: 재고추가 화면에 닫기버튼 및 타이틀 추가, 버튼 액션 함수 이름 수정

Judy-999 committed:
feat: 재고 추가 화면으로 재고 데이터 전달

finnn1 committed:
feat: 재고 추가 화면 오토레이아웃 지정, 재고 수량 레이블에 표시

Judy-999 committed:
feat: Stepper로 재고 수량 레이블 변경

finnn1 committed:
feat: modifyStock 메서드 정의, Stepper로 재고변경 적용

Judy-999 committed:
refactor: 네비게이션 백버튼 제거, Double 확장

finnn1 committed:
fix: 접근제어자 오류 수정

Judy-999 committed:
docs: README에 STEP 3 내용 추가

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        navigationItem.hidesBackButton = true
```
이 부분도 함수로 분리하면 좋을 것 같아요 :)

finnn1 replied:
넵 말씀해주신대로 별도로 함수로 분리해서 처리했습니다 🙏

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    @IBOutlet weak var kiwiStepper: UIStepper!
+    @IBOutlet weak var mangoStepper: UIStepper!
+    
+    var fruitStore = FruitStore(quantity: 0)
```
지금 구조에서는 이전 ViewController에서 여기 `fruitStore` 프로퍼티는 사실상 사용되지 않기 때문에 여기서 생성해주는게 의미가 없습니다!
그래서 이 코드를 보면 처음에 재고가 0부터 시작할 것이라고 오해할 수 있을 것 같아요!
개선해보면 좋을 것 같아요!

finnn1 replied:
아하 그러네요...! 해당 부분에 대해서 고민해보다가 싱글톤 패턴을 활용해서 처리해보았습니다 😊

corykim0829 replied:
그랬군요! 싱글턴을 사용하개 되면 생기는 장단점을 찾아보고 이야기해보면 좋을 것 같아요!

finnn1 replied:
넵 싱글톤의 관한 내용 정리해서 올리겠습니다 ☺️

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,64 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        navigationItem.hidesBackButton = true
+        setStockLabel()
+        setStepper()
```
함수로 깔끔하게 잘 분리하셨습니다 👍

finnn1 replied:
감사합니다 🙏

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -14,6 +14,10 @@ class ViewController: UIViewController {
         updateStock()
     }
+    
+    override func viewWillAppear(_ animated: Bool) {
```
viewWillAppear로 깔끔하게 처리하셨네요 👍

finnn1 replied:
감사합니다..!! 🙏🙏

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -25,17 +29,18 @@ class ViewController: UIViewController {
         let outOfStockAlert = UIAlertController(title: "재료 소진", message: "재료가 모자라요. 재고를 수정할까요?", preferredStyle: .alert)
         let yesAction = UIAlertAction(title: "예", style: .default) { action in
-            guard let stockView = self.storyboard?.instantiateViewController(withIdentifier: "StockViewController") else { return }
+            guard let stockView = self.storyboard?.instantiateViewController(withIdentifier: "StockViewController") as? FruitStockViewController else { return }
```
stockView라고 하면 View라고 오해할 수도 있을 것 같아요! 변수명 수정해주세요!

finnn1 replied:
앗 넵 해당 변수명을 stockViewController로 명시해주었습니다!

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -46,7 +51,7 @@ class ViewController: UIViewController {
     }
     
-    @IBAction func orderJuice(sender: UIButton) {
+    @IBAction func orderJuiceButtonTapped(sender: UIButton) {
```
IBAction 함수명 깔끔합니다 😀

finnn1 replied:
감사합니다 ☺️🙏

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -61,5 +66,11 @@ class ViewController: UIViewController {
     }
     
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == "stockViewSegue" {
```
여기도 `stockViewSegue`에 있는 stockView도 수정해주세요!

finnn1 replied:
과일 재고의 데이터가 있는 클래스를 싱글톤 패턴으로 바꾸었더니 해당 부분이 필요가 없어져 삭제 처리했습니다 🙏

corykim0829 commented on JuiceMaker/JuiceMaker/Model/DoubleExtension.swift:
```diff
@@ -0,0 +1,5 @@
+extension Double {
```
extension을 활용한 부분은 좋은 시도에요!
하지만 모든 Double에 적용되기 때문에, 조심스럽게 사용해야합니다! 가독성과 코드의 유연성을 향상시키기 위해서 사용했다고 하셨는데, 기존 코드에서 
`let stockValue = sender.value` 를 `let stockValue = String(sender.value)`와 같이 처리하면 굳이 extension을 사용하지 않아도 될 것 같아요!

finnn1 replied:
리뷰해주신 내용을 바탕으로 직접 String(format: "%.0f", sender.value) 와 같이 처리해서 Double에 extension을 하지 않도록 처리했습니다! 😊

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -61,5 +66,11 @@ class ViewController: UIViewController {
         updateStock()
```
orderJuiceButtonTapped 함수 안에서 처리하고 있는 부분에서
juiceMaker.make()가 throw하는 error가 하나밖에 없어서 catch 하나만 처리해주면 될 것 같아요! 수정 부탁드립니다 🙏

finnn1 replied:
처음에는 catch JuiceMakerError.outOfStock { } 만 적용할 경우
'Errors thrown from here are not handled because the enclosing catch is not exhaustive'
에러가 발생해서 NSError 를 처리해주기 위해서 catch를 한번 더 사용했었습니다.

그래서 이번에는 리뷰해주신대로 catch 하나로 처리해보았습니다만, 이렇게 진행할 경우 JuiceMakerError와 NSError가 구분이 안될 것 같은데 이대로 진행해도 괜찮은지 궁금해서 질문 드립니다 🤔
이번 프로젝트에서는 NSError가 발생할 일이 없긴 하겠지만, 모든 에러를 처리해주지 않고 구분없이 하나의 catch문으로 묶어서 처리해도 괜찮은 것일까요?

corykim0829 replied:
여기서는 에러를 하나만 던져주는게 확실해서 catch 하나로만 처리하는게 깔끔해보여요!

Catch 안에서 분기처리로 stockError인지 한번 검사하는 것도 좋을 것 같아요 :)

finnn1 replied:
넵 말씀해주신대로 catch 구문 안에서 stockError인지 검사해주는 방식으로 처리했습니다 ☺️
감사합니다 🙏

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -14,6 +14,10 @@ class ViewController: UIViewController {
         super.viewDidLoad()
```
ViewController 클래스명도 역할에 맞게 바꾸면 좋을 것 같아요!

finnn1 replied:
해당 뷰컨트롤러를 역할에 맞도록 JuiceMakerViewController로 변경했습니다 😊🙏

corykim0829 reviewed (changes requested):
핀, 주디 수고 많으셨습니다 😀

### 브랜치 관리
이제 보니 PR을 보낸 브랜치가 핀과 주디의 메인 브랜치네요! 매 스텝마다, 스텝별로 브랜치를 파서 보내는게 좋아요! 다음 미션 때에는 스텝 브랜치로 관리해서 진행해보세요!

---

## 궁금한 점 🥲

### 1. 레이블과 스태퍼 설정 간소화

IBOutlet 들도 outlet collection으로 만들어서 관리 및 사용할 수 있어요! 참고해서 전체적인 로직을 개선해보면 좋을 것 같아요 :)

### 2. FruitStore를 사용한 재고 변경

이 부분은 클래스의 특성을 잘 이용해서 처리하셨다고 생각했어요!
근데 저도 좀 더 고민해보도록 하겠습니다!

---

### 이슈

재고 부족 AlertController가 present 될 때, 아래와 같은 경고 문구가 디버그창에 나옵니다! 해당 경고가 나오지 않도록 수정부탁드려요!

<img width="659" alt="image" src="https://user-images.githubusercontent.com/34022757/166955220-6ee960b2-ecb3-44c3-ab63-9aa4dbb3f52e.png">

finnn1 committed:
refactor: IBOutlet 객체들을 컬렉션 형태로 묶어서 초기화

Judy-999 committed:
refactor: 싱글톤 적용, 함수 로직 변경 및 이름 변경

- FruitStore를 싱글톤 클래스로 만들어 데이더 전달과 재고 관리 방법 변경
- FruitStockViewController에서 뒤로가기 버튼 숨기는 것을 함수로 분리
- 뷰 컨드롤러 변수 이름을 "StockViewController"로 변경
- DoubleExtension 삭제
- ViewController 클래스 이름을 JuiceMakerViewController로 변경
- fix :  재고 부족 알림과 주문 완료 알림이 동시에 뜨지 않도록 수정

corykim0829 commented on JuiceMaker/JuiceMaker/Controller/FruitStockViewController.swift:
```diff
@@ -0,0 +1,45 @@
+
+class FruitStockViewController: UIViewController {
+    
+    @IBOutlet var stockLabelCollection: [UILabel]!
+    @IBOutlet var stepperCollection: [UIStepper]!
```
배열 변수명은 collection 이름을 직접 사용하지 않는게 좋습니다.
labelArray, labelList 이런식으로 사용하지 않고 복수형태로 사용하는게 좋습니다! 참고하세요!

finnn1 replied:
리뷰 해주신 내용을 바탕으로 collection을 List로 변경했습니다 🙏🙏

finnn1 replied:
앗! 남겨주신 내용을 Array나 List로 사용하라는 뜻으로 잘못 이해했네요 😢
다시 Labels나 Steppers 와 같은 복수형태로 변경했습니다 😊

corykim0829 reviewed (changes requested):
수고 많으셨습니다! 👍👍

먼저 싱글톤 처리 관련해서 조금 더 알아보고 같이 이야기 해보면 좋을 것 같습니다!

또한 커밋을 조금 더 잘게 쪼개어서 커밋하면 좋을 것 같아요.
하나의 수정 내용애 대해서 하나의 커밋만 남겨주는게 좋습니다!

finnn1 committed:
refactor: IBOutlet Collection 변수명 변경

Judy-999 committed:
refactor: 재고 부족 에러 분기 처리

Judy-999 commented:
Cory! 리뷰 감사합니다 😊 리뷰해주신 내용을 바탕으로 수정해봤습니다!

## 커밋 단위
수정하는게 부분적으로 있어서 조금씩 수정한다는게 한 번에 해버렸네요..😅 하나의 수정으로 분리해서 커밋하기 명심하겠습니다!
<br>

## 싱글톤 사용
`FruitStore`에서 재고 관리를 하기 때문에 하나의 재고 관리를 하는 것이 의미에 부합하다고 생각하여 싱글톤을 적용해봤습니다. 또한 전역 인스턴스로 사용할 수 있어 뷰 컨트롤러 사이에 따로 데이터 전달이 필요하지 않다는 이점이 있습니다.
물론 전역으로 접근한다는 것이 이점이면서도 동시에 여러 곳에서 접근하기 때문에 문제가 발생하면 추적하기 어렵다는 단점이 있습니다. 이번 프로젝트는 많이 복잡하지 않아 관리할 수 있지만 나중에 사용하게 되면 유의하겠습니다!

finnn1 committed:
refactor: IBOutlet Collection 변수명 복수형으로 변경

corykim0829 reviewed (approved):
수고 많으셨습니다! 👏

싱글톤에 관해서는 [포스트](https://babbab2.tistory.com/66) 참고하셔서 내용 보충하시면 더 좋을 것 같아요! 싱글톤은 수정과 테스트에 걸림돌이 된다는 점 항상 유의하시면 좋을 것 같습니다!

수고 많으셨습니다!!
